src/storage/payloads_storage.cpp

```diff
@@ -3,35 +3,167 @@
 // Distributed under the MIT software license, see the accompanying
 // file LICENSE or http://www.opensource.org/licenses/mit-license.php.
 
+#include <veriblock/algorithm.hpp>
+#include <veriblock/blockchain/alt_block_tree.hpp>
 #include <veriblock/storage/payloads_storage.hpp>
 
 namespace altintegration {
 
-template <>
-PayloadsRepository<ATV>& PayloadsStorage::getRepo() {
-  return *_repoAtv;
+bool PayloadsStorage::getValidity(Slice<const uint8_t> containingBlock,
+                                  Slice<const uint8_t> payloadId) {
+  auto key = makeKey(containingBlock, payloadId);
+  auto it = _cgValidity.find(key);
+  if (it == _cgValidity.end()) {
+    // we don't know if this payload is invalid, so assume it is valid
+    return true;
+  }
+
+  return it->second;
+}
+
+Repository& PayloadsStorage::getRepo() { return *repo_; }
+
+void PayloadsStorage::setValidity(Slice<const uint8_t> containingBlock,
+                                  Slice<const uint8_t> payloadId,
+                                  bool validity) {
+  auto key = makeKey(containingBlock, payloadId);
+  if (!validity) {
+    _cgValidity[key] = validity;
+    return;
+  }
+
+  auto it = _cgValidity.find(key);
+  if (it != _cgValidity.end()) {
+    // this saves some memory, because we assume that
+    // anything that is not in this map is valid by default
+    _cgValidity.erase(it);
+  }
+
+  // do nothing. any entry that is not in this map is valid by default
 }
-template <>
-const PayloadsRepository<ATV>& PayloadsStorage::getRepo() const {
-  return *_repoAtv;
+
+const std::set<AltBlock::hash_t>& PayloadsStorage::getContainingAltBlocks(
+    const std::vector<uint8_t>& payloadId) {
+  return payload_in_alt[payloadId];
```

Comment:
I do not like that we create empty records every time we call this method.

Reply:
this is intentional. otherwise, we would have to return a copy of set

Reply:
What's wrong with having a copy of set? Do you want to modify it after calling get? I think it should be modifiable via some method. And you have a const ref anyway.